src/utils/DirectHub.cxx

```diff
@@ -0,0 +1,943 @@
+            {
+                // We still have unused data in the current buffer. We have to
+                // segment that and send it to the hub.
+                return call_head_segmenter();
+            }
+            if (buf_.free())
+            {
+                // We still have space in the current buffer. We can read more
+                // data into that space.
+                return do_some_read();
+            }
+            else
+            {
+                /// @todo consider not resetting here, but allowing an empty
+                /// but linked DataBuffer* start the next chain.
+                buf_.reset();
+                return alloc_for_read();
+            }
+        }
+
+        /// Current buffer that we are filling.
+        LinkedDataBufferPtr buf_;
+        /// Barrier notifiable to keep track of the buffer's contents.
+        BarrierNotifiable *bufferNotifiable_;
+        /// Output of the last segmenter call.
+        ssize_t segmentSize_;
+        /// 1 if we got the send callback inline from the read_done.
+        uint16_t inlineCall_ : 1;
+        /// 1 if the run callback actually happened inline.
+        uint16_t sendComplete_ : 1;
+        /// Pool of BarrierNotifiables that limit the amount of inflight bytes
+        /// we have.
+        AsyncNotifiableBlock pendingLimiterPool_ {
+            (unsigned)config_directhub_port_max_incoming_packets()};
+        /// Helper object for Select.
+        StateFlowSelectHelper helper_ {this};
+        /// Pointer to the owninng port.
+        DirectHubPortSelect *parent_;
+        /// Implementation (and state) of the business logic that segments
+        /// incoming bytes into messages that shall be given to the hub.
+        std::unique_ptr<MessageSegmenter> segmenter_;
+    } readFlow_;
+
+    friend class DirectHubReadFlow;
+
+public:
+    DirectHubPortSelect(DirectHubInterface<uint8_t[]> *hub, int fd,
+        std::unique_ptr<MessageSegmenter> segmenter,
+        Notifiable *on_error = nullptr)
+        : StateFlowBase(hub->get_service())
+        , readFlow_(this, std::move(segmenter))
+        , readFlowPending_(1)
+        , writeFlowPending_(1)
+        , hub_(hub)
+        , fd_(fd)
+        , onError_(on_error)
+    {
+#ifdef __WINNT__
+        unsigned long par = 1;
+        ioctlsocket(fd_, FIONBIO, &par);
+#else
+        ::fcntl(fd, F_SETFL, O_RDWR | O_NONBLOCK);
+#endif
+
+        // Sets the initial state of the write flow to the stage where we read
+        // the next entry from the queue.
+        wait_and_call(STATE(read_queue));
+        notRunning_ = 1;
+
+        hub_->register_port(this);
+        readFlow_.start();
+        LOG(VERBOSE, "%p create fd %d", this, fd_);
+    }
+
+    ~DirectHubPortSelect()
+    {
+    }
+
+    /// Synchronous output routine called by the hub.
+    void send(MessageAccessor<uint8_t[]> *msg) override
+    {
+        if (fd_ < 0)
+        {
+            // Port already closed. Ignore data to send.
```

Comment:
return?

Reply:
Good catch, thanks! I was seeing some crashes on disconnect, and this might have been the cause of them.